crates/sol-macro/src/expand/contract.rs

```diff
@@ -120,64 +120,75 @@ fn expand_call_like_enum(
     assert_eq!(variants.len(), types.len());
     let name_s = name.to_string();
     let count = variants.len();
-    let min_data_len = min_data_len.min(4);
     quote! {
         #(#attrs)*
         pub enum #name {
             #(#variants(#types),)*
         }
 
-        // TODO: Implement these functions using traits?
         #[automatically_derived]
-        impl #name {
-            /// The number of variants.
-            pub const COUNT: usize = #count;
-
-            // no decode_raw is possible because we need the selector to know which variant to
-            // decode into
-
-            /// ABI-decodes the given data into one of the variants of `self`.
-            pub fn decode(data: &[u8], validate: bool) -> ::alloy_sol_types::Result<Self> {
-                if data.len() >= #min_data_len {
-                    // TODO: Replace with `data.split_array_ref` once it's stable
-                    let (selector, data) = data.split_at(4);
-                    let selector: &[u8; 4] =
-                        ::core::convert::TryInto::try_into(selector).expect("unreachable");
-                    match *selector {
-                        #(<#types as ::alloy_sol_types::#trt>::SELECTOR => {
-                            return <#types as ::alloy_sol_types::#trt>::decode_raw(data, validate)
-                                .map(Self::#variants)
-                        })*
-                        _ => {}
-                    }
+        impl ::alloy_sol_types::SolCalls for #name {
```

Comment:
can we also implement a fallible downcast for each type? `fn as_#variant(&self) -> Option<&#variant_ty>`

Reply:
sure

Reply:
strum does provide `fn is_#variant(&self) -> bool` tho 🤔

Reply:
I've added them anyway, `From`, `TryInto`, `is_*`, and `as_*{,_mut}`. Still not sure about this, maybe we conditionally add them with attributes later (i'm still tinkering with the `#[sol(...)]` attributes)